### PR TITLE
Email signal

### PR DESCRIPTION
**Pull request by Habib-Urr-Rehman**
Please Review

### CODE REVIEW

File: address_book/address_book/urls.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+Including another URLconf
+    1. Import the include() function: from django.urls import include, path
+    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
+"""
+from django.contrib import admin
+from django.urls import path,include
+
+urlpatterns = [
+    path('admin/', admin.site.urls),
+    path('', include('contacts.urls')),
```
Comment (hussain021) [inline address_book/address_book/urls.py]: In your next application please add '/project-name" to this path for experimenting.

File: address_book/contacts/forms.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    class Meta:
+        model = Contact
+        fields = ['first_name', 'last_name', 'email', 'phone_number', 'address']
+
+    def clean_phone_number(self):
+        phone_number = self.cleaned_data['phone_number']
+        try:
+            # Check if any contact with this phone number exists
+            contacts = Contact.objects.filter(phone_number=phone_number)
```
Comment (hussain021) [inline address_book/contacts/forms.py:18]: As pointed out in the last PR, please use get_object_or_404 here or add .first() for this filtering.

File: address_book/contacts/models.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from django.db import models
+
+class Contact(models.Model):
```
Comment (hussain021) [inline address_book/contacts/models.py]: Can we use the built-in AbstractUser or BaseAbstractUser model here?

File: address_book/address_book/urls.py (hunk 1 above)
Comment (Habib-Urr-Rehman) [reply address_book/address_book/urls.py]: Ok sure I will do

File: address_book/contacts/models.py (hunk 3 above)
Comment (Habib-Urr-Rehman) [reply address_book/contacts/models.py]: No, `Abstract User` and `AbstractBaseUser` in Django are meant for managing user authentication, not for storing contact information like names, email, phone numbers, and addresses. In this project we don't need any authentication

File: address_book/contacts/forms.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from django import forms
+from django.core.validators import RegexValidator
+from contacts.models import Contact
+from django.shortcuts import get_object_or_404
+
+class ContactForm(forms.ModelForm):
+    phone_number = forms.CharField(max_length=15, validators=[
+        RegexValidator(r'^\d{1,15}$', message="Phone number must be numeric.")
```
Comment (hussain021) [inline address_book/contacts/forms.py]: Please also allow "+" in the start, update the regex.

File: address_book/contacts/views.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            email_content = f"Name: {name}\nPhone: {phone}\nAddress: {address}\nEmail: {email}"
+
+            # Send the email
+            send_mail(
+                'New Contact Added',
+                email_content,
+                settings.EMAIL_HOST_USER,  # Sender's email address
+                ['2021cs665@student.uet.edu.pk'],  # List of recipients
+            )
+
```
Comment (hussain021) [inline address_book/contacts/views.py]: Remove these extra lines.

File: address_book/contacts/views.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            form.save()
+
+            name = form.cleaned_data['first_name'] + ' ' + form.cleaned_data['last_name']
+            phone = form.cleaned_data['phone_number']
+            address = form.cleaned_data['address']
+            email = form.cleaned_data['email']
+
+            email_content = f"Name: {name}\nPhone: {phone}\nAddress: {address}\nEmail: {email}"
+
+            # Send the email
```
Comment (hussain021) [inline address_book/contacts/views.py]: Please add the link from where did you take this code?

File: address_book/contacts/views.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    return render(request, 'contact_list.html',{'contacts':contacts})
+
+
+def add_contact(request):
+    if request.method == 'POST':
+        form = ContactForm(request.POST)
+        if form.is_valid():
+            form.save()
+
+            name = form.cleaned_data['first_name'] + ' ' + form.cleaned_data['last_name']
```
Comment (hussain021) [inline address_book/contacts/views.py]: use fstring here.

File: address_book/contacts/views.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+from django.shortcuts import render, redirect
```
Comment (hussain021) [inline address_book/contacts/views.py]: Please use proper conventions for imports as per pep8.

**Review by hussain021 — CHANGES_REQUESTED** [review]
Please work on the requested changes.

File: address_book/contacts/forms.py (hunk 4 above)
Comment (Habib-Urr-Rehman) [reply address_book/contacts/forms.py]: yes updated

File: address_book/contacts/views.py (hunk 5 above)
Comment (Habib-Urr-Rehman) [reply address_book/contacts/views.py]: ok sure

File: address_book/contacts/views.py (hunk 6 above)
Comment (Habib-Urr-Rehman) [reply address_book/contacts/views.py]: I have added  reference in changed file

File: address_book/contacts/views.py (hunk 7 above)
Comment (Habib-Urr-Rehman) [reply address_book/contacts/views.py]: updated

File: address_book/contacts/views.py (hunk 8 above)
Comment (Habib-Urr-Rehman) [reply address_book/contacts/views.py]: updated

File: address_book/contacts/views.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+            address = form.cleaned_data['address']
+            email = form.cleaned_data['email']
+            email_content = f"Name: {name}\nPhone: {phone}\nAddress: {address}\nEmail: {email}"
+
+            # Send the email         # ref.https://docs.djangoproject.com/en/5.0/topics/email/
+            send_mail(
+                'New Contact Added',
+                email_content,
+                settings.EMAIL_HOST_USER,  # Sender's email address
+                ['2021cs665@student.uet.edu.pk'],  # List of recipients
```
Comment (hussain021) [inline address_book/contacts/views.py]: Add this in a constant or better yet .env file so it can be changed without making any code changes.

**Review by hussain021 — CHANGES_REQUESTED** [review]
Please add one more change.

**Review by hussain021 — APPROVED** [review]
LGTM! Lets merge.

**Pull request merged**